vignettes/custom_epiworkflows.Rmd

```diff
@@ -0,0 +1,554 @@
+above in the postprocessor of `four_week_ahead$epiworkflow`.
+Creating the layers is a similar process, except with frosting instead of a
+`recipe`[^2]:
+
+```{r make_frosting}
+four_week_layers <- frosting() |>
+  layer_predict() |>
+  layer_residual_quantiles(quantile_levels = c(0.1, 0.25, 0.5, 0.75, 0.9)) |>
+  layer_add_forecast_date() |>
+  layer_add_target_date() |>
+  layer_threshold()
+```
+
+
+Most layers will work for any engine or steps; `layer_predict()` you will want
+to call in every case.
+There are a couple of layers, however, which depend on whether the engine used predicts quantiles or point estimates.
+
+The layers that are only supported by point estimate engines (such as
+`linear_reg()`) are
+
+- `layer_residual_quantiles()`: the preferred method of generating quantiles for
+  non-quantile models, it uses the error residuals of the engine. 
+  This will work for most parsnip engines.
+- `layer_predictive_distn()`: alternate method of generating quantiles, it uses
+  an approximate parametric distribution. This will work for linear regression
+  specifically.
+  
+TODO check this
+  
+On the other hand, the layers that are only supported by quantile estimating
+engines (such as `quantile_reg()`) are
+
+- `layer_quantile_distn()`: adds the specified quantiles.
+  If they differ from the ones actually fit, they will be interpolated and/or
+  extrapolated.
+- `layer_point_from_distn()`: this adds the median quantile as a point estimate,
+  and should be called after `layer_quantile_distn()` if called at all. 
+
+## Fitting an `epi_workflow()`
+
+Given that we now have a recipe and some layers, we need to assemble the workflow:
+```{r workflow_building}
+four_week_workflow <- epi_workflow(
+  four_week_recipe,
+  linear_reg(),
+  four_week_layers
+)
+```
+
+And fit it to recreate `four_week_ahead$epi_workflow`
+```{r workflow_fitting}
+fit_workflow <- four_week_workflow |> fit(training_data)
+```
+
+## Predicting
+
+To do a prediction, we need to first narrow the dataset down to the relevant
+datapoints:
+
+```{r grab_data}
+relevant_data <- get_test_data(
+  four_week_recipe,
+  training_data
+)
+```
+
+With a fit workflow and test data in hand, we can actually make our predictions:
+
+```{r workflow_pred}
+fit_workflow |> predict(relevant_data)
+```
+
+Note that if we had simply plugged `training_data` into `predict()` we still get
+predictions:
+
+```{r workflow_pred_training}
+fit_workflow |> predict(training_data)
+```
+
+The resulting tibble is 800 rows long, however.
+This produces forecasts for not just the actual `forecast_date`, but for every
+day in the dataset it has enough data to actually make a prediction.
+To narrow this down, we could filter to rows where the `time_value` matches the `forecast_date`:
+
+```{r workflow_pred_training_filter}
+fit_workflow |>
+  predict(training_data) |>
+  filter(time_value == forecast_date)
+```
+
+This can be useful for cases where `get_test_data()` doesn't pull sufficient
+data.
```

Comment:
issue: This sounds suspicious. If `get_test_data` decided there wasn't sufficient data, it sounds like the `predict` -> `filter` approach is doing something wrong (predicting with insufficient data).

Reply:
I assumed that the `predict` -> `filter` approach would return the same predictions for `forecast_date`, but it sounds like not.

Reply:
It does produce the same forecasts when it produces forecasts, it's just that sometimes `get_test_data` won't produce a forecast. `get_test_data` is just dumb and won't necessarily grab enough data if the recipe is sufficiently complicated. It's a hack that we need to rework.

Reply:
Modified slightly in https://github.com/cmu-delphi/epipredict/pull/431/commits/d218ed696cf67ceef012301fbfb905cd9d68d8cb